recipes-bsp/firmware-imx/firmware-imx_8.6.bb

```diff
@@ -0,0 +1,118 @@
+        install -m 0644 ${WORKDIR}/epdc-firmware.service ${D}${systemd_system_unitdir}
+        install -m 0644 ${WORKDIR}/regulatory-firmware.service ${D}${systemd_system_unitdir}
+        install -m 0644 ${WORKDIR}/hdmi-firmware.service ${D}${systemd_system_unitdir}
+    fi
+
+    cd firmware
+    for d in *; do
+        case $d in
+        ddr|seco)
+            # These folders are for i.MX 8 and are included in the boot image via imx-boot
+            bbnote Excluding folder $d
+            ;;
+        *)
+            cp -rfv $d ${D}${base_libdir}/firmware
+            ;;
+        esac
+    done
+    cd -
+
+    # Install SDMA Firmware: sdma-imx6q.bin & sdma-imx7d.bin into lib/firmware/imx/sdma
+    install -d ${D}${base_libdir}/firmware/imx/sdma
+    mv ${D}${base_libdir}/firmware/sdma/sdma-imx6q.bin ${D}${base_libdir}/firmware/imx/sdma
+    mv ${D}${base_libdir}/firmware/sdma/sdma-imx7d.bin ${D}${base_libdir}/firmware/imx/sdma
+
+    mv ${D}${base_libdir}/firmware/epdc/ ${D}${base_libdir}/firmware/imx/epdc/
+    mv ${D}${base_libdir}/firmware/imx/epdc/epdc_ED060XH2C1.fw.nonrestricted ${D}${base_libdir}/firmware/imx/epdc/epdc_ED060XH2C1.fw
+
+    mv ${D}${base_libdir}/firmware/easrc/ ${D}${base_libdir}/firmware/imx/easrc/
+
+    # Install HDMI Firmware: hdmitxfw.bin, hdmirxfw.bin & dpfw.bin into lib/firmware/imx/hdmi
+    install -d ${D}${base_libdir}/firmware/imx/hdmi
+    mv ${D}${base_libdir}/firmware/hdmi/cadence/hdmitxfw.bin ${D}${base_libdir}/firmware/imx/hdmi
+    mv ${D}${base_libdir}/firmware/hdmi/cadence/hdmirxfw.bin ${D}${base_libdir}/firmware/imx/hdmi
+    mv ${D}${base_libdir}/firmware/hdmi/cadence/dpfw.bin ${D}${base_libdir}/firmware/imx/hdmi
+
+    find ${D}${base_libdir}/firmware -type f -exec chmod 644 '{}' ';'
+    find ${D}${base_libdir}/firmware -type f -exec chown root:root '{}' ';'
+
+    # Remove files not going to be installed
+    find ${D}${base_libdir}/firmware/ -name '*.mk' -exec rm '{}' ';'
+    rm -rf ${D}${base_libdir}/firmware/hdmi
+}
+
+python populate_packages_prepend() {
+    vpudir = bb.data.expand('${base_libdir}/firmware/vpu', d)
+    do_split_packages(d, vpudir, '^vpu_fw_([^_]*).*\.bin',
+                      output_pattern='firmware-imx-vpu-%s',
+                      description='Freescale IMX Firmware %s',
+                      extra_depends='',
+                      prepend=True)
+
+    sdmadir = bb.data.expand('${base_libdir}/firmware/sdma', d)
+    do_split_packages(d, sdmadir, '^sdma-([^-]*).*\.bin',
+                      output_pattern='firmware-imx-sdma-%s',
+                      description='Freescale IMX Firmware %s',
+                      extra_depends='',
+                      prepend=True)
+}
+
+ALLOW_EMPTY_${PN} = "1"
+
+PACKAGES_DYNAMIC = "${PN}-vpu-* ${PN}-sdma-*"
+
+PACKAGES =+ "${PN}-epdc ${PN}-scfw ${PN}-sdma ${PN}-easrc ${PN}-regulatory ${PN}-hdmi"
+
+RDEPENDS_${PN}-epdc = "bash"
+RDEPENDS_${PN}-sdma = "bash"
+RDEPENDS_${PN}-regulatory = "bash"
+RDEPENDS_${PN}-hdmi = "bash"
```

Comment:
How about those who would like to build GPLv3-less images?

`bash` cannot be used as a shell on those, therefore what this recipe would do it to lockdown licensing terms for images, since those FW blobs are quite essential.

Could those additional scripts be re-worked to be shell-neutral and *bashism-free*?

Reply:
Hmm, good point on this one, I'll look into. Thanks!

> @panfir I dropped few comments and questions regarding this PR.
> 
> In general, can you use `.sh` extensions for scripts? This would be easier to identify on the target then.
> 
> -- andrey

Well, it's good to keep it that way, the `shebang` is enough.
It is not meant to be explicitly identified on the target as a regular script, it is run as `systemd.service` at boot.
I'm not sure any rules about it but I haven't seen the use of extensions to the scripts used by `systemd` services.
I general, I think this rule is used:
.. /`script_name`
.. /`script_name.service`

Reply:
> How about those who would like to build GPLv3-less images?
> 
> `bash` cannot be used as a shell on those, therefore what this recipe would do it to lockdown licensing terms for images, since those FW blobs are quite essential.

Hi Andrey, I just checked with our trusted advisors ..
As far as Licensing goes, it depends on what these are doing with bash.  If bash is being compiled into anything, that is a problem.  If we are modifying bash and re-distributing it, that could be a problem.  If it is just being required because we are using bash to run in and we are just pulling it from the external source, it is not a problem that bash is GPL-v3.

> 
> Could those additional scripts be re-worked to be shell-neutral and _bashism-free_?

Reply:
> > How about those who would like to build GPLv3-less images?
> > `bash` cannot be used as a shell on those, therefore what this recipe would do it to lockdown licensing terms for images, since those FW blobs are quite essential.
> 
> Hi Andrey, I just checked with our trusted advisors ..
> As far as Licensing goes, it depends on what these are doing with bash. If bash is being compiled into anything, that is a problem. If we are modifying bash and re-distributing it, that could be a problem. If it is just being required because we are using bash to run in and we are just pulling it from the external source, it is not a problem that bash is GPL-v3.
> 

I have a different point of view regarding usage of GPLv3 components is final target images, namely the so-called `Tivoization clause`. You might want to get back to the legal department with exactly this question, but my understanding of that clause is:

Once GPLv3-licensed component is included in any firmware image which is signed by vendor and provided as a binary to any party, then this party can request from the firmware vendor mechanisms and means to re-create and re-flash that firmware image, which automatically means that firmware vendor should make signing keys and facilities public.

Usually, this is is not a concern as most of devices built are purposed in the way they could be updated without having those signatures in place.

But some HW (and firmware) vendors are placing restrictions on their design (security, vendor lock-ins, etc) and therefore opting-out from GPLv3 components in their images. This could be done in Yocto by blacklisting GPLv3 and excluding all components from the image.

What this commit would effectively do - it introduces the core component required for HW to operate and at the same time enforcing the license for the dependent component (`bash` in this case) to be set to GPLv3, automatically making final FW file non-GPLv3-free, thus making all those HW and FW vendors above to turn away from this platform, or seek alternatives like taking an older version of this layer.

I'm not really sure if this was an intention here with introducing `bash` dependency.

As I said: I'm not a layer, therefore please treat above as my personal opinion and a guideline, rather than a legal advise or any statement of some sort. You might would want to check the above with the Legal department.

*"If in doubt - call the Lawyer..."* 😉 

-- andrey

Reply:
It will not be the case, we'll get rid of bash from the f/w loading scripts. Interestingly, the so-called `Tivoization clause` and Yes, GPLv3 is a problem, we want to avoid it as much as possible. Thanks!